pkg/util/log/log.go

```diff
@@ -452,99 +448,155 @@ func formatErrorc(message string, context ...interface{}) error {
 // scrubAndLogFunc, and treating the variadic args as the message.
 func log(logLevel seelog.LogLevel, bufferFunc func(), scrubAndLogFunc func(string), v ...interface{}) {
 	l := logger.Load()
-	if l != nil && l.inner != nil && l.shouldLog(logLevel) {
+
+	if l == nil {
+		addLogToBuffer(bufferFunc)
+		return
+	}
+
+	l.l.Lock() //// If we change the logic here we will want to change in addLogToBuffer too
+	defer l.l.Unlock()
+
+	if l.inner != nil && l.shouldLog(logLevel) {
 		s := BuildLogEntry(v...)
-		l.l.Lock()
-		defer l.l.Unlock()
+
 		scrubAndLogFunc(s)
-	} else if l == nil || l.inner == nil {
+	} else if l.inner == nil {
 		addLogToBuffer(bufferFunc)
 	}
 }
 
 func logWithError(logLevel seelog.LogLevel, bufferFunc func(), scrubAndLogFunc func(string) error, fallbackStderr bool, v ...interface{}) error {
 	l := logger.Load()
-	if l != nil && l.inner != nil && l.shouldLog(logLevel) {
+
+	if l == nil {
+		addLogToBuffer(bufferFunc)
+		err := formatError(v...)
+		if fallbackStderr && (l == nil) {
+			fmt.Fprintf(os.Stderr, "%s: %s\n", logLevel.String(), err.Error())
+		}
+		return err
+	}
+
+	l.l.Lock()
+	defer l.l.Unlock()
```

Comment:
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏You keep holding the lock even after the `if` block, when you call `formatError` and `fmt.Sprintf`, we probably don't want to do that.
You only need the lock held when reading `l.inner` and calling `scrubAndLogFunc`

Reply:
There is the same issue every time you `defer` unlocking outside of the `if` block